### ISSUE TITLE
Rule: decode Base64 

### ISSUE THREAD

**Issue by azu** [secretlint/secretlint#19]
base64 is not encrypted logic.
We should decode base64 text and check it.

**Comment by azu** [comment]
we need to provide helper?

**Comment by azu** [comment]
It will be `weak-enctrypto` rule.

**Comment by azu** [comment]
Enctrypted token may be catched as false-positive.

For example, following token is enctrypted with password. but some rule detect it as raw token.

```
    AWS_SECRET_ACCESS_KEY: R6WCNd0erkAuZyD+Fe8C4IwjpYQtniKqnHAzE7Pk5fPrBRpwOnuYxA==
```

**Comment by azu** [comment]
Add little tips https://github.com/secretlint/secretlint/blob/master/docs/secretlint-rule.md

**Comment by azu** [comment]
https://github.blog/changelog/2025-02-14-secret-scanning-detects-base64-encoded-github-tokens/
GitHub Supported encoded Base64 detection.

```
/(Z2hw|Z2hv|Z2h1|Z2hz|Z2hy)Xz[A-Za-z0-9+\/]{48}={0,2}/
```

**Comment by konard** [comment]
### A production incident where this gap cost a live credential

We hit this in [link-assistant/hive-mind#2156](https://github.com/link-assistant/hive-mind/issues/2156) last week, and the data may be useful here since the thread has been open a while.

A GitHub OAuth token (`gho_`) leaked through a published log and was revoked by GitHub's scanner. It was never in the log in plaintext: GHCR's token endpoint echoes the credential it is given back to the caller, base64-encoded, inside a JSON body. Secretlint was running on that publication path with `@secretlint/secretlint-rule-preset-recommend` and reported nothing.

**Reproduction** (recommended preset only, synthetic token, `@secretlint/core` 13.0.2):

```js
import { lintSource } from '@secretlint/core';
import { creator } from '@secretlint/secretlint-rule-preset-recommend';

const token = `gho_${'A'.repeat(12)}SyntheticNotReal${'9'.repeat(8)}`;
const encoded = Buffer.from(token).toString('base64');
const config = { rules: [{ id: '@secretlint/secretlint-rule-preset-recommend', rule: creator }] };

for (const [label, content] of [
  ['plain', `token: ${token}`],
  ['base64', `{"token":"${encoded}"}`],
  ['base64 in escaped JSON', `{"stdout":"{\\"token\\":\\"${encoded}\\"}"}`],
  ['base64 wrapped at 20 cols', encoded.replace(/(.{20})/g, '$1\n')],
]) {
  const { messages } = await lintSource({
    source: { filePath: '/x.log', content, contentType: 'text' },
    options: { config, maskSecrets: false },
  });
  console.log(label.padEnd(28), messages.length);
}
```

```
plain                        1
base64                       0
base64 in escaped JSON       0
base64 wrapped at 20 cols    0
```

Run against the actual 16.9 MB incident log — which contains six base64-encoded credentials — the recommended preset reports **0 findings in 391 ms**.

### On the regex in [this comment](https://github.com/secretlint/secretlint/issues/19#issuecomment-2660645177)

We measured `/(Z2hw|Z2hv|Z2h1|Z2hz|Z2hy)Xz[A-Za-z0-9+\/]{48}={0,2}/` before considering it, and it is much narrower than it looks — **10 of 186** token/alignment combinations in the shape we actually leaked. Two limits stack:

1. **Byte alignment.** It only matches a token beginning at a 3-byte boundary of the blob. A token inside a JSON body starts wherever the prefix puts it, and `{"token":"` is 10 characters, so two of the three alignments can never match.
2. **The literal `Xz` pins the token's fifth character.** `_` is `0x5F`, so the second sextet is `0x30 | (c5 >> 4)`, which equals `z` only for `c5` in `0x30`–`0x3F` — the ten digits out of the sixty-two characters a token body uses. That is ~16% of tokens even at alignment 0.

```
alignment 0, by 5th character of the token:  matched 10/62 (0123456789)
byte alignment (a token that matches at 0):  offset 0 true, offset 1 false, offset 2 false
embedded in a JSON body, across alignments:  matched 10/186
```

Script: [`issue-2156-base64-token-regex-coverage.mjs`](https://github.com/link-assistant/hive-mind/blob/f99077245f3a611c3b72840e641a7dc221b9bffd/experiments/issue-2156-base64-token-regex-coverage.mjs).

### On the false-positive objection in [this comment](https://github.com/secretlint/secretlint/issues/19#issuecomment-592312206)

@azu's point that an *encrypted* value can decode into something a rule matches is correct, and it was the main design constraint for us too. What resolved it was **not reporting on decoded text directly**:

1. find bounded decodable runs (base64 at all three alignments, base64url, hex, percent, byte-escape, HTML entity);
2. decode each run and scan it **separately** — scanning a joined document lets a rule match across a boundary and blame an innocent run;
3. mask inside the decoded payload, re-encode, and **verify the round trip** — substitute only if decoding the rebuilt run returns exactly the intended output, otherwise redact the whole run.

A false positive then costs one masked run and never a corrupted document, which makes the tradeoff acceptable in a way that reporting decoded matches is not. It also preserves partial masking through the encoding: `{"token":"gho_…"}` stays valid base64 and decodes to `{"token":"gho…999"}`, so a human can still correlate the value across a log.

Line wrapping needs handling to be useful in practice — `base64(1)` and MIME fold at 76 columns, PEM at 64 — so a folded blob contains no contiguous single-line run at all. We scan contiguous groups of base64-only lines and offer both readings (with and without the partial lines at either end), keeping whichever decodes.

Implementation, if useful as a reference: [`src/encoded-credential-detection.lib.mjs`](https://github.com/link-assistant/hive-mind/blob/f99077245f3a611c3b72840e641a7dc221b9bffd/src/encoded-credential-detection.lib.mjs), with the write-up in [this case study](https://github.com/link-assistant/hive-mind/blob/f99077245f3a611c3b72840e641a7dc221b9bffd/docs/case-studies/issue-2156/README.md). Happy to extract any of it into a PR here if a `weak-encrypto`-style rule or a shared decode helper is the direction you want.
